components/calendar/src/chinese.rs

```diff
@@ -0,0 +1,1044 @@
+/// calendar corresponding to the inception of the calendar, marked as 2637 BCE (ISO: -2636), and negative
+/// years marking Chinese years before February 15, 2637 BCE.
+///
+/// Because the Chinese calendar does not traditionally count years, era codes are not used in this calendar;
+/// internally, a single era code "era" is used, but this is never checked, so era code input can be anything.
+/// This Chinese calendar implementation does support a related ISO year, which marks the ISO year in which a
+/// Chinese year begins, and a cyclic year corresponding to the year in the 60 year cycle as described above.
+///
+/// For more information, suggested reading materials include:
+/// * _Calendrical Calculations_ by Reingold & Dershowitz
+/// * _The Mathematics of the Chinese Calendar_ by Helmer Aslaksen (https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.139.9311&rep=rep1&type=pdf)
+/// * Wikipedia: https://en.wikipedia.org/wiki/Chinese_calendar
+///
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Chinese;
+
+/// The inner date type used for representing [`Date`]s of [`Chinese`]. See [`Date`] and [`Chinese`] for more details.
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq, PartialOrd, Ord)]
+pub struct ChineseDateInner(ArithmeticDate<Chinese>);
+
+impl CalendarArithmetic for Chinese {
+    fn month_days(year: i32, month: u8) -> u8 {
+        let mid_year = Self::fixed_mid_year_from_year(year);
+        let new_year = Chinese::chinese_new_year_on_or_before_fixed_date(mid_year);
+        let approx = new_year + ((month - 1) as i64 * 29);
+        let prev_new_moon = Chinese::chinese_new_moon_before((approx + 15).as_moment());
+        let next_new_moon = Chinese::chinese_new_moon_on_or_after((approx + 15).as_moment());
+        (next_new_moon - prev_new_moon) as u8
+
+        // let mid_year = Self::fixed_mid_year_from_year(year);
+        // let new_year = Chinese::chinese_new_year_on_or_before_fixed_date(mid_year);
+        // let mut cur_month: u8 = 1;
+        // let mut cur_rata_die = new_year;
+        // let mut iters: u8 = 0;
+        // let max_iters: u8 = 13;
+        // while cur_month < month && iters < max_iters {
+        //     cur_rata_die = Chinese::chinese_new_moon_on_or_after((cur_rata_die + 1).as_moment());
+        //     cur_month += 1;
+        //     iters += 1;
+        // }
+        // debug_assert!(iters < max_iters, "Unexpectedly large number of iterations");
+        // (Chinese::chinese_new_moon_on_or_after((cur_rata_die + 1).as_moment()) - cur_rata_die) as u8
+    }
+
+    fn months_for_every_year(year: i32) -> u8 {
+        if Self::is_leap_year(year) {
+            13
+        } else {
+            12
+        }
+    }
+
+    fn is_leap_year(year: i32) -> bool {
+        let mid_year = Self::fixed_mid_year_from_year(year);
+        Self::fixed_date_is_in_leap_year(mid_year)
+    }
+
+    fn last_month_day_in_year(year: i32) -> (u8, u8) {
```

Comment:
issue for @sffc: we should probably switch the `max_date` usages to using constants and then include `debug_assert`s that their value is correct. These last_month_day methods are going to get expensive and we're only using them to compute exactly one value